src/core/handshaker/security/pipelined_secure_endpoint.cc

```diff
@@ -0,0 +1,938 @@
+            *impl->pending_output_buffer_ =
+                std::move(*impl->unprotected_data_buffer_);
+            impl->unprotected_data_buffer_.reset();
+            on_read = std::move(impl->on_read_);
+            lock.Release();
+            impl->event_engine_->Run([on_read = std::move(on_read)]() mutable {
+              on_read(absl::OkStatus());
+            });
+          } else {
+            // We're waiting on the next read to read this data. We're done
+            // for now.
+            impl->waiting_for_transport_read_ = true;
+            lock.Release();
+            break;
+          }
+        }
+      }
+    }
+
+    static void StartAsyncRead(grpc_core::RefCountedPtr<Impl> impl,
+                               ReadArgs args) {
+      grpc_core::ReleasableMutexLock lock(&impl->shutdown_read_mu_);
+      if (impl->wrapped_ep_ == nullptr) {
+        lock.Release();
+        FailReads(std::move(impl),
+                  absl::CancelledError("secure endpoint shutdown"));
+        return;
+      }
+      const bool read_finished_immediately = impl->wrapped_ep_->Read(
+          [impl = impl->Ref()](absl::Status status) mutable {
+            bool should_unprotect = false;
+            {
+              grpc_core::MutexLock lock(&impl->read_queue_mu_);
+              should_unprotect = impl->ShouldContinueUnprotect();
+              if (!status.ok()) {
+                // We rely on ContinueUnprotect to fail the read if the
+                // status is not ok, since we might currently have an
+                // unprotect in progress and we want to return that data
+                // before we fail future reads.
+                impl->unprotecting_ = status;
+              } else {
+                impl->frame_protector_.TraceOp(
+                    "Read",
+                    impl->staging_protected_data_buffer_->c_slice_buffer());
+                impl->MoveStagingIntoProtectedBuffer();
+              }
+            }
+            // We now have data to unprotect so we should continue unprotecting
+            // if not currently doing so.
+            if (should_unprotect) {
+              ContinueUnprotect(std::move(impl));
+            }
+          },
+          impl->staging_protected_data_buffer_.get(), std::move(args));
+      if (read_finished_immediately) {
+        lock.Release();
+        bool should_unprotect = false;
+        {
+          grpc_core::MutexLock lock(&impl->read_queue_mu_);
+          impl->frame_protector_.TraceOp(
+              "ReadImm",
+              impl->staging_protected_data_buffer_->c_slice_buffer());
+          impl->MoveStagingIntoProtectedBuffer();
+          should_unprotect = impl->ShouldContinueUnprotect();
+        }
+        // We now have data to unprotect so we should continue unprotecting
+        // if not currently doing so.
+        if (should_unprotect) {
+          ContinueUnprotect(std::move(impl));
+        }
+      }
+    }
+
+    grpc_core::Mutex write_queue_mu_;
```

Comment:
It'd be worth annotating mutexes with ABSL_ACQUIRED_BEFORE/AFTER for all mutexes in this class (if an ordering exists)

Reply:
Done.